### PR TITLE
Kinesis post, part 1

### PR DESCRIPTION
**Pull request by anahitpo**
This is a long one, sorry..:) I highly recommend reviewing in preview, reading the markdown will be tedious.

### CODE REVIEW

File: _posts/2020-05-25-kinesis-streams-part-1.md [hunk 1]
```diff
@@ -0,0 +1,439 @@
...
+
+Each SDK has its own ways of configuring the request timeouts. Node.js SDK allows you to set two distinct types of timeous:
+- `connectTimeout`: a timeout for establishing a **new connection** on a socket; if not explicitly set, this value will default to the value of the `timeout`
+- `timeout`: read timeout for an existing socket, or, in other words, the time between when the request ends and the response is received, **including service and network round-trips**
+
+It's important to understand the differences between the two. It's equally important to know what the default values are and to configure the timeout values according to the expected latencies of the services you are using.
+
+In the case of Kinesis API, the default timeout on an existing socket (the `timeout` configuration) is **2 minutes**. This means that, by default, AWS SDK will wait for the response for the entire two minutes, before it gives up and retries the request. Add to that the default 2 minutes for establishing a new connection and possible retries that follow both of the timeouts, and you get a producer that is stuck for a while waiting for something that might never happen.
```
Comment (phusu) [inline _posts/2020-05-25-kinesis-streams-part-1.md:300]: timeout on an existing socket -> timeout of an existing socket

File: _posts/2020-05-25-kinesis-streams-part-1.md [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+
+# Monitoring a stream
+
+Now when we have been talking about failures for quite some time (because everything fails, all the time, right? :)), it's time to talk about how you can detect failures and monitor your stream's performance. _CloudWatch_ has a [set of metrics](https://docs.aws.amazon.com/streams/latest/dev/monitoring-with-cloudwatch.html) related to _Kinesis Streams_.
+
+The so-called "**basic**" metrics are enabled by default and tell you about the stream in overall. You can also enable the "**enhanced**" metrics, which allow you to monitor the stream on the level of individual shards. The enhanced shard-level metrics can be very useful, especially when you suspect having an "overheated" shard, but they will cost you extra. The basic stream-level metrics are free.
+
+Below are the most important metrics to keep an eye on when writing to a Kinesis stream:
+
+- `IncomingRecords`: The **number** of records **successfully** written to the Kinesis stream. This includes records from both, `PutRecord` and `PutRecods` operations. In case you want to separate the two, each of them has its own metric as well: `PutRecord.Success` and `PutRecords.Records` respectively.
```
Comment (phusu) [inline _posts/2020-05-25-kinesis-streams-part-1.md:361]: Remove comma: This includes records from both `PutRecord` and `PutRecods` operations.

File: _posts/2020-05-25-kinesis-streams-part-1.md [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+
+
+# Conclusions
+
+There seems to be a common misconception that serverless and fully managed mean that the services never fail. But, as we know, everything fails and we should always have a failure in mind when developing our systems. Not to be pessimistic, but to **be prepared**.
+
+Here are some of the key takeaways I hope you find useful. The bonus here is that many of them apply to most of the API calls done through AWS SDK (think of _SQS_, _DynamoDB_, _Kinesis Firehose_, etc.)
+
+### Batch operations and partial failures:
+When using batch operations, such as the `PutRecords` API for _Kinesis Streams_ (and let's face it, it will be most of the time, unless u have a solid reason not to), always allow for **partial failures**. They will happen. To be frank, if you haven't addressed them yet, they are probably happening right now and you might have no idea. And trust me, it's an awkward moment when you realize that. But the moment is better to come sooner rather than later :)
```
Comment (phusu) [inline _posts/2020-05-25-kinesis-streams-part-1.md:405]: u -> you

File: _posts/2020-05-25-kinesis-streams-part-1.md [hunk 4]
```diff
@@ -0,0 +1,439 @@
...
+  - having data retention over 24 h (pricing is per shard per hour)
+  - Enhanced (shard-level) CloudWatch metrics (per shard per metric per month)
+  - using enhanced fan-out
+  - DynamoDB charges when using KCL
+
+
+----
+
+
+And this is it, my friends! If you managed to read this till the end, you should have a solid idea of how _Kinesis Streams_ work and how you can write data to a stream, avoiding some of the most common pitfalls. In **Part 2**, I'll be talking about reading from a stream, in particular using _AWS Lambda_ as a stream consumer.
```
Comment (phusu) [inline _posts/2020-05-25-kinesis-streams-part-1.md:437]: And this is it -> And that's it

File: _posts/2020-05-25-kinesis-streams-part-1.md (hunk 3 above)
Comment (anahitpo) [reply _posts/2020-05-25-kinesis-streams-part-1.md:405]: 🙈

File: _posts/2020-05-25-kinesis-streams-part-1.md (hunk 1 above)
Comment (anahitpo) [reply _posts/2020-05-25-kinesis-streams-part-1.md:300]: I think for would be even better!

File: _posts/2020-05-25-kinesis-streams-part-1.md [hunk 5]
```diff
@@ -0,0 +1,439 @@
...
+   },
+   {
+     ErrorCode: 'ProvisionedThroughputExceededException',
+     ErrorMessage: 'Rate exceeded for shard shardId-00001 in stream someStream under account xxx.'
+   }
+ ]
+}
+```
+
+There is an array of records again, though having a different content this time, and an additional attribute called `FailedErrorCount`. If the value of `FailedErrorCount` is **greater than 0**, this means that you had a **partial failure** of your request. This, in turn, means that part of your records have been written to the stream successfully, while some of them have failed. I guess you could also call it a partial success if you fall on the more optimist side of the spectrum :)
```
[the remark this answers is not in the document]
Comment (anahitpo) [reply _posts/2020-05-25-kinesis-streams-part-1.md:184]: damn, I have no idea where the `FailedErrorCount` came from! 🙈

File: _posts/2020-05-25-kinesis-streams-part-1.md [hunk 6]
```diff
@@ -0,0 +1,439 @@
...
+
+But I digress.
+
+
+
+# Writing to a stream
+
+So, how do you actually write data to a Kinesis stream? Well, you have plenty of options for that!
+
+You can use the [Amazon Kinesis Agent](https://docs.aws.amazon.com/firehose/latest/dev/writing-with-agents.html), which is a stand-alone Java application to stream data from files and is useful to f.ex. stream your server logs to the cloud. The next option you have is the [Amazon Kinesis Producer Library (KPL)](https://docs.aws.amazon.com/streams/latest/dev/developing-producers-with-kpl.html). It provides a higher level of abstraction over the API calls and has several useful features I'll be mentioning shortly.
```
[the remark this answers is not in the document]
Comment (anahitpo) [reply _posts/2020-05-25-kinesis-streams-part-1.md:65]: definitely!

File: _posts/2020-05-25-kinesis-streams-part-1.md [hunk 7]
```diff
@@ -0,0 +1,439 @@
...
+
+![autoscaling](/img/kinesis/autoscaling.png){: .img.centered }
+
+Why not just add an enormous amount of shards to a stream to accommodate all possible spikes in incoming traffic, you might be asking? You could, but that will cost you. In addition to paying for the amount of data being streamed, every open shard is being charged for on an hourly basis.
+
+Though being fully managed, in a sense, _Kinesis Streams_ are still not entirely serverless, while you need to provision the capacity in the form of shards, pay for that provisioned capacity even if it is not fully utilized and manage the scaling yourself. In a way, it is like DynamoDB in its early days. And like in the case with DynamoDB, I do firmly believe the actual autoscaling is coming, at some point, soon. This, of course, might be just my wishful thinking :)
```
[the remark this answers is not in the document]
Comment (anahitpo) [reply _posts/2020-05-25-kinesis-streams-part-1.md:55]: great point!

File: _posts/2020-05-25-kinesis-streams-part-1.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (phusu) [reply _posts/2020-05-25-kinesis-streams-part-1.md:65]: I just pointed out the first usage of "f.ex.", there are a lot more. Find all & replace :)

**Pull request merged**